lib/gitlab/regex.rb

```diff
@@ -44,13 +44,20 @@ def project_path_regex_message
 
 
     def file_name_regex
-      @file_name_regex ||= /\A[a-zA-Z0-9_\-\.]*\z/.freeze
+      @file_name_regex ||= /\A[a-zA-Z0-9_\-\.\/]*\z/.freeze
     end
 
     def file_name_regex_message
-      "can contain only letters, digits, '_', '-' and '.'. "
+      "can contain only letters, digits, '_', '-' and '.'. Separate directories with a '/'. "
     end
 
+    def directory_traversal_regex
+      @directory_traversal_regex ||= /\.{2}/.freeze
```

Comment:
FYI, gitlab_git will resolve `foo/../bar` properly to `/bar`. It might still make sense to keep this check, though.